# WeRateDogs: Wrangle and Analysis

In [1]:
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import timeit

## Gathering

### Manually

In [2]:
#Read the twitter data provided by Udacity
df_twitter = pd.read_csv ('twitter-archive-enhanced.csv')
df_twitter.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### Programatically

In [3]:
#Download of image predictions file with requests library
image_url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(image_url)

with open('image_predictions.tsv', 'wb') as f:
    f.write(response.content)

In [4]:
#Read the image prediction data
df_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
df_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
# Save df_predictions in a csv file
df_predictions.to_csv('image_predictions.csv', index=False)

### API

In [6]:
#Introduction of Twitter's keys
consumer_key = 'Secret-key'
consumer_secret = 'Secret-key'
access_token = 'Secret-key'
access_secret = 'Secret-key'

#Connection with the Twitter's API
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

In [6]:
#Create a list of Id's from the file df_twitter
tweet_id_list = df_twitter['tweet_id'].tolist()

In [27]:
#Get data from the Twitter's API
start = timeit.timeit()

tweet_errors = []
with open('tweet_json.txt', 'w') as f:
    for i in tweet_id_list:       
        try:
            tweet_row = api.get_status(i, tweet_mode = 'extended', wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
            json.dump(tweet_row._json, f)
            f.write('\n')
            
        except Exception as ex:
            print(ex)
            print(f'Error_tweet_id {i}')
            tweet_errors.append(i)
            continue

end = timeit.timeit()
print(end - start)

[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 888202515573088257
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 873697596434513921
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 872668790621863937
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 872261713294495745
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 869988702071779329
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 866816280283807744
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 861769973181624320
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 856602993587888130
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 851953902622658560
[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 845459076796616705
[{'code': 144, 'message': 'No status found with that ID.'}]


Rate limit reached. Sleeping for: 163


[{'code': 144, 'message': 'No status found with that ID.'}]
Error_tweet_id 680055455951884288


Rate limit reached. Sleeping for: 136


-0.0004242000013618963


In [45]:
# Read tweets in json format from tweet_json.txt and convert it to a Dataframe
json_list = []
with open('tweet_json.txt', 'r') as file:
    for row in file:
        try:
            tweet = json.loads(row)
            json_list.append({'tweet_id': tweet['id'], 'retweet_count': tweet['retweet_count'], 
                              'favorite_count': tweet['favorite_count']})
          
        except Exception as ex:
            print(ex)
            continue

df_api = pd.DataFrame(json_list)
df_api.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7455           35310
1  892177421306343426           5535           30551
2  891815181378084864           3661           22994
3  891689557279858688           7625           38592
4  891327558926688256           8221           36878

In [9]:
# Save df_api in a csv file
df_api.to_csv('tweet_json.csv', encoding = 'utf-8', index=False)

## Assesing

### df_twitter

In [24]:
df_twitter

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [56]:
df_twitter.sort_values('name', ascending=False)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
773   776249906839351296                    NaN                  NaN   
1031  745422732645535745                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
1097  736225175608430592                    NaN                  NaN   
1385  700747788515020802                    NaN                  NaN   
...                  ...                    ...                  ...   
1327  705975130514706432                    NaN                  NaN   
1933  674019345211760640                    NaN                  NaN   
938   753294487569522689                    NaN                  NaN   
1021  746726898085036033                    NaN                  NaN   
1035  744995568523612160                    NaN                  NaN   

                      timestamp  \
773   2016-09-15 02:42:54 +0000   
1031  2016-06-22 01:06:43 +0000   
819   2016-08-30 16:11:18 +0000   
1097  2016-05-27 15:58:54 +0000   
1385  2016-02-19 18:24:26 +0000   
...                         ...   
1327  2016-03-05 04:36:02 +0000   
1933  2015-12-08 00:15:09 +0000   
938   2016-07-13 18:26:16 +0000   
1021  2016-06-25 15:29:00 +0000   
1035  2016-06-20 20:49:19 +0000   

                                                 source  \
773   <a href="http://twitter.com/download/iphone" r...   
1031  <a href="http://twitter.com/download/iphone" r...   
819   <a href="http://twitter.com/download/iphone" r...   
1097  <a href="http://twitter.com/download/iphone" r...   
1385  <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1327  <a href="http://twitter.com/download/iphone" r...   
1933  <a href="http://twitter.com/download/iphone" r...   
938   <a href="http://twitter.com/download/iphone" r...   
1021  <a href="http://twitter.com/download/iphone" r...   
1035  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
773   RT @dog_rates: We only rate dogs. Pls stop sen...         7.007478e+17   
1031  We only rate dogs. Pls stop sending in non-can...                  NaN   
819   We only rate dogs. Pls stop sending in non-can...                  NaN   
1097  We only rate dogs. Please stop sending in non-...                  NaN   
1385  We only rate dogs. Pls stop sending in non-can...                  NaN   
...                                                 ...                  ...   
1327  This is Adele. Her tongue flies out of her mou...                  NaN   
1933  This is Acro. You briefly see her out of the c...                  NaN   
938   This is Ace. He's a window washer. One of the ...                  NaN   
1021  Meet Abby. She's incredibly distracting. Just ...                  NaN   
1035  This is Abby. She got her face stuck in a glas...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
773               4.196984e+09  2016-02-19 18:24:26 +0000   
1031                       NaN                        NaN   
819                        NaN                        NaN   
1097                       NaN                        NaN   
1385                       NaN                        NaN   
...                        ...                        ...   
1327                       NaN                        NaN   
1933                       NaN                        NaN   
938                        NaN                        NaN   
1021                       NaN                        NaN   
1035                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
773   https://twitter.com/dog_rates/status/700747788...                11   
1031  https://twitter.com/dog_rates/status/745422732...                 9   
819   https://twitter.com/dog_rates/status/770655142...                11   
1097  https://twitt

In [83]:
df_twitter.name.value_counts()

None       745
a           55
Charlie     12
Cooper      11
Oliver      11
          ... 
Arlen        1
Cecil        1
all          1
Eevee        1
Rhino        1
Name: name, Length: 957, dtype: int64

In [88]:
df_twitter[df_twitter.name == 'a']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
56    881536004380872706                    NaN                  NaN   
649   792913359805018113                    NaN                  NaN   
801   772581559778025472                    NaN                  NaN   
1002  747885874273214464                    NaN                  NaN   
1004  747816857231626240                    NaN                  NaN   
1017  746872823977771008                    NaN                  NaN   
1049  743222593470234624                    NaN                  NaN   
1193  717537687239008257                    NaN                  NaN   
1207  715733265223708672                    NaN                  NaN   
1340  704859558691414016                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1361  703079050210877440                    NaN                  NaN   
1368  702539513671897089                    NaN                  NaN   
1382  700864154249383937                    NaN                  NaN   
1499  692187005137076224                    NaN                  NaN   
1737  679530280114372609                    NaN                  NaN   
1785  677644091929329666                    NaN                  NaN   
1853  675706639471788032                    NaN                  NaN   
1854  675534494439489536                    NaN                  NaN   
1877  675109292475830276                    NaN                  NaN   
1878  675047298674663426                    NaN                  NaN   
1923  674082852460433408                    NaN                  NaN   
1941  673715861853720576                    NaN                  NaN   
1955  673636718965334016                    NaN                  NaN   
1994  672604026190569472                    NaN                  NaN   
2034  671743150407421952                    NaN                  NaN   
2066  671147085991960577                    NaN                  NaN   
2116  670427002554466305                    NaN                  NaN   
2125  670361874861563904                    NaN                  NaN   
2128  670303360680108032                    NaN                  NaN   
2146  669923323644657664                    NaN                  NaN   
2153  669661792646373376                    NaN                  NaN   
2161  669564461267722241                    NaN                  NaN   
2191  668955713004314625                    NaN                  NaN   
2198  668815180734689280                    NaN                  NaN   
2211  668614819948453888                    NaN                  NaN   
2218  668507509523615744                    NaN                  NaN   
2222  668466899341221888                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2304  666983947667116034                    NaN                  NaN   
2311  666781792255496192                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN         

In [89]:
df_twitter[df_twitter.name == 'None']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
5     2017-07-29 00:08:17 +0000   
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2342  2015-11-16 02:38:37 +0000   
2343  2015-11-16 01:59:36 +0000   
2344  2015-11-16 01:52:02 +0000   
2351  2015-11-16 00:24:50 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
5     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2342  <a href="http://twitter.com/download/iphone" r...   
2343  <a href="http://twitter.com/download/iphone" r...   
2344  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
5     Here we have a majestic great white breaching ...                  NaN   
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2342  Here we have a well-established sunblockerspan...                  NaN   
2343  Let's hope this flight isn't Malaysian (lol). ...                  NaN   
2344  Here we have a northern speckled Rhododendron....                  NaN   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
5                          NaN                        NaN   
7                          NaN                        NaN   
12                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2342                       NaN                        NaN   
2343                       NaN                        NaN   
2344                       NaN                        NaN   
2351                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
5     https://twitter.com/dog_rates/status/891087950...                13   
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
24    https://twitt

In [85]:
lista = df_twitter.name.str.match(r"\b[^A-Z]\w+")
df_twitter[lista]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
22    887517139158093824                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
169   859196978902773760                    NaN                  NaN   
193   855459453768019968                    NaN                  NaN   
335   832645525019123713                    NaN                  NaN   
369   828650029636317184                    NaN                  NaN   
542   806219024703037440                    NaN                  NaN   
682   788552643979468800                    NaN                  NaN   
759   778396591732486144                    NaN                  NaN   
773   776249906839351296                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
852   765395769549590528                    NaN                  NaN   
924   755206590534418437                    NaN                  NaN   
988   748977405889503236                    NaN                  NaN   
992   748692773788876800                    NaN                  NaN   
993   748575535303884801                    NaN                  NaN   
1025  746369468511756288                    NaN                  NaN   
1031  745422732645535745                    NaN                  NaN   
1040  744223424764059648                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1071  740214038584557568                    NaN                  NaN   
1095  736392552031657984                    NaN                  NaN   
1097  736225175608430592                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1121  730924654643314689                    NaN                  NaN   
1138  728035342121635841                    NaN                  NaN   
1206  715758151270801409                    NaN                  NaN   
1259  710272297844797440                    NaN                  NaN   
1362  703041949650034688                    NaN                  NaN   
1385  700747788515020802                    NaN                  NaN   
1435  697259378236399616                    NaN                  NaN   
1457  695095422348574720                    NaN                  NaN   
1527  690360449368465409                    NaN                  NaN   
1603  685943807276412928                    NaN                  NaN   
1693  681297372102656000                    NaN                  NaN   
1724  680085611152338944                    NaN                  NaN   
1747  679111216690831360                    NaN                  NaN   
1797  677269281705472000                    NaN                  NaN   
1815  676613908052996102                    NaN                  NaN   
1916  674307341513269249                    NaN                  NaN   
1936  673956914389192708                    NaN                  NaN   
2001  672482722825261057                    NaN                  NaN   
2019  672125275208069120                    NaN                  NaN   
2030  671789708968640512                    NaN                  NaN   
2037  671561002136281088                    NaN                  NaN   
2204  668636665813057536                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   
2326  666411507551481857                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2349  666051853826850816                    NaN                  NaN   

                      timestamp  \
22    2017-07-19 03:

In [90]:
df_twitter[lista].name.value_counts()

the             8
an              7
very            5
one             4
quite           4
just            4
mad             2
actually        2
getting         2
not             2
light           1
officially      1
my              1
by              1
unacceptable    1
infuriating     1
such            1
all             1
life            1
this            1
old             1
incredibly      1
his             1
space           1
Name: name, dtype: int64

In [91]:
df_twitter[df_twitter.name == 'quite']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
118   869988702071779329                    NaN                  NaN   
169   859196978902773760                    NaN                  NaN   
193   855459453768019968                    NaN                  NaN   
2326  666411507551481857                    NaN                  NaN   

                      timestamp  \
118   2017-05-31 18:47:24 +0000   
169   2017-05-02 00:04:57 +0000   
193   2017-04-21 16:33:22 +0000   
2326  2015-11-17 00:24:19 +0000   

                                                 source  \
118   <a href="http://twitter.com/download/iphone" r...   
169   <a href="http://twitter.com/download/iphone" r...   
193   <a href="http://twitter.com/download/iphone" r...   
2326  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
118   RT @dog_rates: We only rate dogs. This is quit...         8.591970e+17   
169   We only rate dogs. This is quite clearly a smo...                  NaN   
193   Guys, we only rate dogs. This is quite clearly...                  NaN   
2326  This is quite the dog. Gets really excited whe...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
118               4.196984e+09  2017-05-02 00:04:57 +0000   
169                        NaN                        NaN   
193                        NaN                        NaN   
2326                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
118   https://twitter.com/dog_rates/status/859196978...                12   
169   https://twitter.com/dog_rates/status/859196978...                12   
193   https://twitter.com/dog_rates/status/855459453...                12   
2326  https://twitter.com/dog_rates/status/666411507...                 2   

      rating_denominator   name doggo floofer pupper puppo  
118                   10  quite  None    None   None  None  
169                   10  quite  None    None   None  None  
193                   10  quite  None    None   None  None  
2326                  10  quite  None    None   None  None

In [92]:
df_twitter[df_twitter.name == 'very']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
773   776249906839351296                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
1031  745422732645535745                    NaN                  NaN   
1097  736225175608430592                    NaN                  NaN   
1385  700747788515020802                    NaN                  NaN   

                      timestamp  \
773   2016-09-15 02:42:54 +0000   
819   2016-08-30 16:11:18 +0000   
1031  2016-06-22 01:06:43 +0000   
1097  2016-05-27 15:58:54 +0000   
1385  2016-02-19 18:24:26 +0000   

                                                 source  \
773   <a href="http://twitter.com/download/iphone" r...   
819   <a href="http://twitter.com/download/iphone" r...   
1031  <a href="http://twitter.com/download/iphone" r...   
1097  <a href="http://twitter.com/download/iphone" r...   
1385  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
773   RT @dog_rates: We only rate dogs. Pls stop sen...         7.007478e+17   
819   We only rate dogs. Pls stop sending in non-can...                  NaN   
1031  We only rate dogs. Pls stop sending in non-can...                  NaN   
1097  We only rate dogs. Please stop sending in non-...                  NaN   
1385  We only rate dogs. Pls stop sending in non-can...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
773               4.196984e+09  2016-02-19 18:24:26 +0000   
819                        NaN                        NaN   
1031                       NaN                        NaN   
1097                       NaN                        NaN   
1385                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
773   https://twitter.com/dog_rates/status/700747788...                11   
819   https://twitter.com/dog_rates/status/770655142...                11   
1031  https://twitter.com/dog_rates/status/745422732...                 9   
1097  https://twitter.com/dog_rates/status/736225175...                10   
1385  https://twitter.com/dog_rates/status/700747788...                11   

      rating_denominator  name doggo floofer pupper puppo  
773                   10  very  None    None   None  None  
819                   10  very  None    None   None  None  
1031                  10  very  None    None   None  None  
1097                  10  very  None    None   None  None  
1385                  10  very  None    None   None  None

In [93]:
df_twitter[df_twitter.name == 'an']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
759   778396591732486144                    NaN                  NaN   
1025  746369468511756288                    NaN                  NaN   
1362  703041949650034688                    NaN                  NaN   
2204  668636665813057536                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2349  666051853826850816                    NaN                  NaN   

                      timestamp  \
759   2016-09-21 00:53:04 +0000   
1025  2016-06-24 15:48:42 +0000   
1362  2016-02-26 02:20:37 +0000   
2204  2015-11-23 03:46:18 +0000   
2333  2015-11-16 19:31:45 +0000   
2335  2015-11-16 16:11:11 +0000   
2349  2015-11-16 00:35:11 +0000   

                                                 source  \
759   <a href="http://twitter.com/download/iphone" r...   
1025  <a href="http://twitter.com/download/iphone" r...   
1362  <a href="http://twitter.com/download/iphone" r...   
2204  <a href="http://twitter.com/download/iphone" r...   
2333  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   
2349  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
759   RT @dog_rates: This is an East African Chalupa...         7.030419e+17   
1025  This is an Iraqi Speed Kangaroo. It is not a d...                  NaN   
1362  This is an East African Chalupa Seal. We only ...                  NaN   
2204  This is an Irish Rigatoni terrier named Berta....                  NaN   
2333  This is an extremely rare horned Parthenon. No...                  NaN   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  NaN   
2349  This is an odd dog. Hard on the outside but lo...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
759               4.196984e+09  2016-02-26 02:20:37 +0000   
1025                       NaN                        NaN   
1362                       NaN                        NaN   
2204                       NaN                        NaN   
2333                       NaN                        NaN   
2335                       NaN                        NaN   
2349                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
759   https://twitter.com/dog_rates/status/703041949...                10   
1025  https://twitter.com/dog_rates/status/746369468...                 9   
1362  https://twitter.com/dog_rates/status/703041949...                10   
2204  https://twitter.com/dog_rates/status/668636665...                10   
2333  https://twitter.com/dog_rates/status/666337882...                 9   
2335  https://twitter.com/dog_rates/status/666287406...                 1   
2349  https://twitter.com/dog_rates/status/666051853...                 2   

      rating_denominator name doggo floofer pupper puppo  
759                   10   an  None    None   None  None  
1025                  10   an  None    None   None  None  
1362                  10   an  None    None   None  None  
2204                  10   an  None    None   None  None  
2333                  10   an  None    None   None  None  
2335                   2   an  None    None   None  None  
2349                  10   an  None    None   None  None

In [94]:
df_twitter[df_twitter.name == 'incredibly']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
542  806219024703037440                    NaN                  NaN   

                     timestamp  \
542  2016-12-06 19:29:28 +0000   

                                                source  \
542  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
542  We only rate dogs. Please stop sending in non-...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
542                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
542  https://twitter.com/dog_rates/status/806219024...                11   

     rating_denominator        name doggo floofer pupper puppo  
542                  10  incredibly  None    None   None  None

## Quality
### df_twitter
- Line 852 name Zoey insted of my
- Lines with in_reply not null are replies
- Lines with retweeted status not null are retweets
- Dog with name incredibly is not a dog
- Dogs with name quite 118, 169, 193 are not dogs
- Dogs with name very are not dogs
- Lines 801, 1025, 1193, 1207, 1737, 1854, 1877 are not dogs
- Line 992 name is Quizno, not his
- Names a, an, quite, None,...[list] should be changed to NaN
- timestamp column is a string
- There should be an extra column showing the actual rating without differences in the denominators

## Tideness

- **df_twitter**: Columns doggo, ploofer, pupper, puppo should be only one
- **df_twitter, df_predictions and df_api** should be merged as similar contents


In [43]:
df_twitter[df_twitter.name == 'very']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
773   776249906839351296                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
1031  745422732645535745                    NaN                  NaN   
1097  736225175608430592                    NaN                  NaN   
1385  700747788515020802                    NaN                  NaN   

                      timestamp  \
773   2016-09-15 02:42:54 +0000   
819   2016-08-30 16:11:18 +0000   
1031  2016-06-22 01:06:43 +0000   
1097  2016-05-27 15:58:54 +0000   
1385  2016-02-19 18:24:26 +0000   

                                                 source  \
773   <a href="http://twitter.com/download/iphone" r...   
819   <a href="http://twitter.com/download/iphone" r...   
1031  <a href="http://twitter.com/download/iphone" r...   
1097  <a href="http://twitter.com/download/iphone" r...   
1385  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
773   RT @dog_rates: We only rate dogs. Pls stop sen...         7.007478e+17   
819   We only rate dogs. Pls stop sending in non-can...                  NaN   
1031  We only rate dogs. Pls stop sending in non-can...                  NaN   
1097  We only rate dogs. Please stop sending in non-...                  NaN   
1385  We only rate dogs. Pls stop sending in non-can...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
773               4.196984e+09  2016-02-19 18:24:26 +0000   
819                        NaN                        NaN   
1031                       NaN                        NaN   
1097                       NaN                        NaN   
1385                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
773   https://twitter.com/dog_rates/status/700747788...                11   
819   https://twitter.com/dog_rates/status/770655142...                11   
1031  https://twitter.com/dog_rates/status/745422732...                 9   
1097  https://twitter.com/dog_rates/status/736225175...                10   
1385  https://twitter.com/dog_rates/status/700747788...                11   

      rating_denominator  name doggo floofer pupper puppo  
773                   10  very  None    None   None  None  
819                   10  very  None    None   None  None  
1031                  10  very  None    None   None  None  
1097                  10  very  None    None   None  None  
1385                  10  very  None    None   None  None

In [8]:
df_twitter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [9]:
df_twitter.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [11]:
df_twitter[df_twitter['retweeted_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
...                  ...                    ...                  ...   
1023  746521445350707200                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
1242  711998809858043904                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                 source  \
19    <a href="http://twitter.com/download/iphone" r...   
32    <a href="http://twitter.com/download/iphone" r...   
36    <a href="http://twitter.com/download/iphone" r...   
68    <a href="http://twitter.com/download/iphone" r...   
73    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1023  <a href="http://twitter.com/download/iphone" r...   
1043  <a href="http://twitter.com/download/iphone" r...   
1242  <a href="http://twitter.com/download/iphone" r...   
2259  <a href="http://twitter.com" rel="nofollow">Tw...   
2260  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
19    RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36    RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68    RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73    RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   
...                                                 ...                  ...   
1023  RT @dog_rates: This is Shaggy. He knows exactl...         6.678667e+17   
1043  RT @dog_rates: Extremely intelligent dog here....         6.671383e+17   
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. ...         7.119983e+17   
2259  RT @dogratingrating: Exceptional talent. Origi...         6.675487e+17   
2260  RT @dogratingrating: Unoriginal idea. Blatant ...         6.675484e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
19                4.196984e+09  2017-07-19 00:47:34 +0000   
32                1.960740e+07  2017-07-15 02:44:07 +0000   
36                4.196984e+09  2017-02-12 01:04:29 +0000   
68                4.196984e+09  2017-06-23 01:10:23 +0000   
73                4.196984e+09  2017-06-23 16:00:04 +0000   
...                        ...                        ...   
1023              4.196984e+09  2015-11-21 00:46:50 +0000   
1043              4.196984e+09  2015-11-19 00:32:12 +0000   
1242              7.832140e+05  2016-03-21 19:29:52 +0000   
2259              4.296832e+09  2015-11-20 03:43:06 +0000   
2260              4.296832e+09  2015-11-20 03:41:59 +0000   

                                          expanded_urls  rating_numerator  \
19    https://twitter.com/dog_rates/status/887473957...                13   
32    https://twitter.com/dog_rates/status/886053434...                12   
36    https://twitter.com/dog_rates/status/830583320...                13   
68    https://twitt

In [12]:
df_twitter[df_twitter['in_reply_to_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
...                  ...                    ...                  ...   
2038  671550332464455680           6.715449e+17         4.196984e+09   
2149  669684865554620416           6.693544e+17         4.196984e+09   
2169  669353438988365824           6.678065e+17         4.196984e+09   
2189  668967877119254528           6.689207e+17         2.143566e+07   
2298  667070482143944705           6.670655e+17         4.196984e+09   

                      timestamp  \
30    2017-07-15 16:51:35 +0000   
55    2017-07-02 21:58:53 +0000   
64    2017-06-27 12:14:36 +0000   
113   2017-06-02 19:38:25 +0000   
148   2017-05-13 16:15:35 +0000   
...                         ...   
2038  2015-12-01 04:44:10 +0000   
2149  2015-11-26 01:11:28 +0000   
2169  2015-11-25 03:14:30 +0000   
2189  2015-11-24 01:42:25 +0000   
2298  2015-11-18 20:02:51 +0000   

                                                 source  \
30    <a href="http://twitter.com/download/iphone" r...   
55    <a href="http://twitter.com/download/iphone" r...   
64    <a href="http://twitter.com/download/iphone" r...   
113   <a href="http://twitter.com/download/iphone" r...   
148   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2038  <a href="http://twitter.com/download/iphone" r...   
2149  <a href="http://twitter.com/download/iphone" r...   
2169  <a href="http://twitter.com/download/iphone" r...   
2189  <a href="http://twitter.com/download/iphone" r...   
2298  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
30    @NonWhiteHat @MayhewMayhem omg hello tanner yo...                  NaN   
55    @roushfenway These are good dogs but 17/10 is ...                  NaN   
64                      @RealKentMurphy 14/10 confirmed                  NaN   
113   @ComplicitOwl @ShopWeRateDogs &gt;10/10 is res...                  NaN   
148   @Jack_Septic_Eye I'd need a few more pics to p...                  NaN   
...                                                 ...                  ...   
2038  After 22 minutes of careful deliberation this ...                  NaN   
2149  After countless hours of research and hundreds...                  NaN   
2169  This is Tessa. She is also very pleased after ...                  NaN   
2189                     12/10 good shit Bubka\n@wane15                  NaN   
2298  After much debate this dog is being upgraded t...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
30                         NaN                        NaN   
55                         NaN                        NaN   
64                         NaN                        NaN   
113                        NaN                        NaN   
148                        NaN                        NaN   
...                        ...                        ...   
2038                       NaN                        NaN   
2149                       NaN                        NaN   
2169                       NaN                        NaN   
2189                       NaN                        NaN   
2298                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
30                                                  NaN                12   
55                                                  NaN                17   
64                                                  NaN                14   
113                

In [54]:
sum(df_twitter.duplicated())

0

### Predictions

In [23]:
df_predictions

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [17]:
df_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [18]:
df_predictions.p1.value_counts()

golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
pug                    57
                     ... 
convertible             1
sunglasses              1
cliff                   1
ibex                    1
mortarboard             1
Name: p1, Length: 378, dtype: int64

In [53]:
sum(df_predictions.duplicated())

0

### API

In [46]:
df_api

tweet_id  retweet_count  favorite_count
0     892420643555336193           7455           35310
1     892177421306343426           5535           30551
2     891815181378084864           3661           22994
3     891689557279858688           7625           38592
4     891327558926688256           8221           36878
...                  ...            ...             ...
2326  666049248165822465             39              95
2327  666044226329800704            124             263
2328  666033412701032449             39             108
2329  666029285002620928             41             118
2330  666020888022790149            447            2357

[2331 rows x 3 columns]

In [47]:
df_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2331 non-null   int64
 1   retweet_count   2331 non-null   int64
 2   favorite_count  2331 non-null   int64
dtypes: int64(3)
memory usage: 54.8 KB


In [52]:
sum(df_api.duplicated())

0